ReText/singleapplication.py

```diff
@@ -0,0 +1,167 @@
+# This file is part of ReText
+# Copyright: 2012-2015 Dmitry Shachnev
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import functools
+import struct
+from PyQt5.QtCore import pyqtSignal
+from PyQt5.QtCore import QObject
+from PyQt5.QtCore import QSharedMemory
+from PyQt5.QtCore import QSystemSemaphore
+from PyQt5.QtNetwork import QLocalServer
+from PyQt5.QtNetwork import QLocalSocket
+
+class SingleApplication(QObject):
+	"""
+	SingleApplication is a class that wrap around the single application
+	framework.
+	"""
+
+	# Modes
+	(
+		# Server mode indicated that we started the first application
+		Server,
+		# Client mode means that another App already started, we should just
+		# exit or send some message to that App.
+		Client,
+	) = range(0, 2)
+
+	# Signals
+
+	receivedMessage = pyqtSignal(bytes)
+
+	def __init__(self, name, parent=None):
+		QObject.__init__(self, parent)
+		self._name = name
+		self._mode = self.Server
+		self._sharedMemory = QSharedMemory(self._name)
```

Comment:
Each user on the system should be able to run this application, but currently the code forces one instance per system.
It would be easy enough to include the username here, but that would still make it very easy for a malicious user to perform a denial of service attack on others. Can you think of a way around this?